### PR TITLE
Add save_video capability to pipeline

### PR DESCRIPTION
**Pull request by cat101**

### CODE REVIEW

File: landingai/vision_pipeline.py [hunk 1]
```diff
@@ -204,6 +210,49 @@ def save_image(self, filename_prefix: str, image_src: str = "") -> "FrameSet":
             c += 1
         return self
 
+    def save_video(
```
Comment (AsiaCao) [inline landingai/vision_pipeline.py:213]: 1. please add unit tests for it, cover a few cases where `video_fps` is none versus not none, `video_length_sec` is none versus not none.
2. how about create a low level API first, i.e. in a dedicated video module, then create wrapper in vision pipeline.

File: landingai/vision_pipeline.py [hunk 2]
```diff
@@ -204,6 +210,49 @@ def save_image(self, filename_prefix: str, image_src: str = "") -> "FrameSet":
...
+        """Save the FrameSet as an mp4 video file
+
+        Parameters
+        ----------
+        video_file_path : str
+            Path and filename with extension of the video file
+        video_fps : Union[int, None], optional
+            Either the frames per second or the video length should be provided to assemble the video
+        video_length_sec : Union[int, None], optional
+            Either the frames per second or the video length should be provided to assemble the video
```
Comment (AsiaCao) [inline landingai/vision_pipeline.py:229]: please also document the behavior when `video_fps` or `video_length_sec` is none

Comment (cat101) [reply landingai/vision_pipeline.py]: Done

File: landingai/vision_pipeline.py [hunk 3]
```diff
@@ -45,9 +45,15 @@ def run_predict(self, predictor: Predictor) -> "Frame":
         self.predictions = predictor.predict(np.asarray(self.image))  # type: ignore
         return self
 
-    def to_numpy_array(self) -> np.ndarray:
-        """Return a numpy array using GRB color encoding (used by OpenCV)"""
-        return np.asarray(self.image)
+    def to_numpy_array(self, image_src: str = "") -> np.ndarray:
+        """Return a numpy array using GRB color encoding (used by OpenCV)
```
[the remark this answers is not in the document]
Comment (cat101) [reply landingai/vision_pipeline.py:49]: Yes. Thanks!

File: landingai/vision_pipeline.py (hunk 1 above)
Comment (cat101) [reply landingai/vision_pipeline.py:213]: Unit test added. Layering I prefer to delay it until we need it

Comment (AsiaCao) [reply landingai/vision_pipeline.py:213]: Wouldn't it be better if we reverse the priority?
Prioritize lower level function over the higher layer, we can defer introducing the higher layer if the bandwidth is limited. The foundation (lower level function) is less likely to be changed than higher level functions.

This could be generalized as a rule of thumb for future function/cases, maybe worth an in-person discussion.

File: landingai/vision_pipeline.py [hunk 4]
```diff
@@ -89,7 +95,7 @@ def from_array(cls, array: np.ndarray, is_BGR: bool = True) -> "FrameSet":
         array : np.ndarray
             Image
         is_BGR : bool, optional
-            Assume OpenCV's BGR color space? Defaults to True
+            Assume OpenCV's BGR channel ordering? Defaults to True
```
Comment (AsiaCao) [inline landingai/vision_pipeline.py:98]: If we want the SDK to be more friendly, maybe a better way is to handle RGB conversion at the source level: image file, video, camera etc. For simple cases like image file reading, leverage PIL is a better approach.

This parameter is a little problematic and I'd just remove it, since
1) it's hard to tell if a numpy array is BGR or RGB as the layout information is lost at this point. So the user may not even know it, especially for novice CV developers.
2) aside from BGR, and RGB, there are other common layouts like grayscale, rgba, etc. It's probably better to use an Enum with a dedicated function to support it instead of a flag.

(btw, the naming should be `is_bgr` instead (snake case), and the default value/behavior should probably be RGB)

File: landingai/vision_pipeline.py [hunk 5]
```diff
@@ -204,6 +210,64 @@ def save_image(self, filename_prefix: str, image_src: str = "") -> "FrameSet":
...
+        """
+        if not video_file_path.lower().endswith(".mp4"):
+            raise NotImplementedError("Only .mp4 is supported")
+        total_frames = len(self.frames)
+        if total_frames == 0:
+            return self
+
+        if video_fps is not None and video_length_sec is not None:
+            raise ValueError(
+                "The fps and length arguments cannot be set at the same time"
```
Comment (AsiaCao) [inline landingai/vision_pipeline.py:252]:
```suggestion
                "The 'video_fps' and 'video_length_sec' arguments cannot be set at the same time"
```

File: landingai/vision_pipeline.py [hunk 6]
```diff
@@ -204,6 +210,64 @@ def save_image(self, filename_prefix: str, image_src: str = "") -> "FrameSet":
...
+                    break
+                frs.extend(frame)
+            frs.save_video("sample_images/test.mp4",video_fps=fps)
+        ```
+
+        Parameters
+        ----------
+        video_file_path : str
+            Path and filename with extension of the video file
+        video_fps : Union[int, None], optional
```
Comment (AsiaCao) [inline landingai/vision_pipeline.py:237]:
```suggestion
        video_fps : Optional[int]
```

File: landingai/vision_pipeline.py [hunk 7]
```diff
@@ -204,6 +210,64 @@ def save_image(self, filename_prefix: str, image_src: str = "") -> "FrameSet":
...
+            Either the frames per second or the video length should be provided to assemble the video. if none of the two are provided, the method will try to set a "reasonable" value.
+        video_length_sec : Union[int, None], optional
```
Comment (AsiaCao) [inline landingai/vision_pipeline.py:239]:
```suggestion
        video_length_sec : Optional[int]
```

btw, should this be float instead?

File: landingai/vision_pipeline.py [hunk 8]
```diff
@@ -204,6 +210,64 @@ def save_image(self, filename_prefix: str, image_src: str = "") -> "FrameSet":
...
+                frs.extend(frame)
+            frs.save_video("sample_images/test.mp4",video_fps=fps)
+        ```
+
+        Parameters
+        ----------
+        video_file_path : str
+            Path and filename with extension of the video file
+        video_fps : Union[int, None], optional
+            Either the frames per second or the video length should be provided to assemble the video. if none of the two are provided, the method will try to set a "reasonable" value.
```
Comment (AsiaCao) [inline landingai/vision_pipeline.py:238]:
```suggestion
            The number of frames per second for the output video file.
            Either the `video_fps` or `video_length_sec ` should be provided to assemble the video. if none of the two are provided, the method will try to set a "reasonable" value.
```

File: landingai/vision_pipeline.py [hunk 9]
```diff
@@ -204,6 +210,64 @@ def save_image(self, filename_prefix: str, image_src: str = "") -> "FrameSet":
...
+        video_length_sec : Union[int, None], optional
+            Either the frames per second or the video length should be provided to assemble the video. if none of the two are provided, the method will try to set a "reasonable" value.
```
Comment (AsiaCao) [inline landingai/vision_pipeline.py:240]:
```suggestion
            The total number of seconds for the output video file. This value will be capped at the number of frames available in this Frameset.
            Either the `video_fps` or `video_length_sec` should be provided to assemble the video. if none of the two are provided, the method will try to set a "reasonable" value.
            If both `video_fps` and `video_length_sec` are provided, `video_length_sec` takes precedence over `video_fps`.
```

Comment (AsiaCao) [reply landingai/vision_pipeline.py:240]: it'd be beneficial to explain:
1. how the "reasonable" value is calculated, the users may get want to know the behavior to decide how to use it, i.e. wether the user should provide a default value.
2. the behavior when both arguments are provided.

File: landingai/vision_pipeline.py [hunk 10]
```diff
@@ -204,6 +210,64 @@ def save_image(self, filename_prefix: str, image_src: str = "") -> "FrameSet":
...
+        if total_frames == 0:
+            return self
+
+        if video_fps is not None and video_length_sec is not None:
+            raise ValueError(
+                "The fps and length arguments cannot be set at the same time"
+            )
+
+        # Try to tune FPS based on parameters or pick a reasonable number
+        if video_length_sec is not None and video_length_sec <= total_frames:
```
Comment (AsiaCao) [inline landingai/vision_pipeline.py:256]: WDYT if we remove the `video_length_sec <= total_frames` check and make `video_fps` a float? 
(btw, this behavior could cause confusion as the code is not respecting the input value, and it's not documented.)

File: landingai/vision_pipeline.py (hunk 1 above)
Comment (cat101) [reply landingai/vision_pipeline.py:213]: Add the topic to our discussion list

File: landingai/vision_pipeline.py (hunk 4 above)
Comment (cat101) [reply landingai/vision_pipeline.py:98]: Note that the pipeline uses only PIL on its interfaces for the same reason that you mentioned (or at least I tried). In this case since this is an input/interface function we need it since the array may be coming from other sources

You have a good point about the other layouts... in fact we had an issue `overlay_bboxes` because it could not handle RGBA.... I have added a todo 

snake case comment:done

File: landingai/vision_pipeline.py (hunk 6 above)
Comment (cat101) [reply landingai/vision_pipeline.py:237]: Thanks

File: landingai/vision_pipeline.py (hunk 7 above)
Comment (cat101) [reply landingai/vision_pipeline.py:239]: Sounds good

File: landingai/vision_pipeline.py (hunk 10 above)
Comment (cat101) [reply landingai/vision_pipeline.py:258]: I added a comment to explain the reasoning (which is tricky):

_Try to tune FPS based on parameters or pick a reasonable number. The goal is to produce a video that last a a couple of seconds even when there are few frames. OpenCV will silently fail and not create a file if the resulting fps is less than 1_

File: landingai/vision_pipeline.py [hunk 11]
```diff
@@ -204,6 +210,66 @@ def save_image(self, filename_prefix: str, image_src: str = "") -> "FrameSet":
...
+        if total_frames == 0:
+            return self
+
+        if video_fps is not None and video_length_sec is not None:
+            raise ValueError(
+                "The 'video_fps' and 'video_length_sec' arguments cannot be set at the same time"
+            )
+
+        # Try to tune FPS based on parameters or pick a reasonable number. The goal is to produce a video that last a a couple of seconds even when there are few frames. OpenCV will silently fail and not create a file if the resulting fps is less than 1
+        if video_length_sec is not None and video_length_sec <= total_frames:
```
Comment (AsiaCao) [inline landingai/vision_pipeline.py:258]: got it, if FPS can't be lower than 1, how about the API automatically caps it at 1 fps? see below

```suggestion
        if video_length_sec > total_frames:
              video_length_sec = total_frames        
        if video_length_sec is not None and video_length_sec <= total_frames:
```

it is worth a documentation to avoid confusion and clarify the behavior.

Comment (cat101) [reply landingai/vision_pipeline.py:258]: This is already happening in line 260.... I try to make the default FPS 2 (for faster video) but if there is a single frame then FPS = 1....

Comment (AsiaCao) [reply landingai/vision_pipeline.py:258]: i think line 260 doesn't cover the case when both `video_length_sec` and `video_fps` is not None.
in addition, the current behavior is a bit confusing too. let's discuss it in our standup.

Comment (cat101) [reply landingai/vision_pipeline.py:259]: Added to dicsussion list
